backend/matching-service/src/controllers/matchingController.ts

```diff
@@ -26,3 +28,86 @@ export async function insertMatching(matchingInfo: MatchingInfo) {
     throw error;
   }
 }
+
+function findIntersection(arr1: string[], arr2: string[]) {
+  // If one user has no preference, take the preferences of the other user.
+  if (!arr1) return arr2;
+  if (!arr2) return arr1;
+
+  // If they both have preferences, take only what is common between them.
+  return arr1.filter((item) => arr2.includes(item));
+}
+
+export async function findMatch(matchingInfo: MatchingInfo) {
+  const difficulty_level_query: FilterQuery<typeof Matching>[] =
+    matchingInfo.difficulty_level
+      ? [
+          {
+            $or: [
+              {
+                difficulty_levels: {
+                  $elemMatch: { $in: matchingInfo.difficulty_level },
+                },
+              },
+              {
+                difficulty_levels: {
+                  $size: 0,
+                },
+              },
+            ],
+          },
+        ]
+      : [];
+
+  const topics_query: FilterQuery<typeof Matching>[] = matchingInfo.topics
+    ? [
+        {
+          $or: [
+            {
+              categories: {
+                $elemMatch: { $in: matchingInfo.topics },
+              },
+            },
+            {
+              categories: {
+                $size: 0,
+              },
+            },
+          ],
+        },
+      ]
+    : [];
+
+  const potentialMatch = await Matching.findOne({
```

Comment:
initially i considered if there are multiple matches, whether we shd choose the most compatible or the earliest, but i think the earliest makes more sense since waiting time is prolly more impt